### ISSUE TITLE
How to inference a BiSeNetV2 model (onnx) that was trained with custom data?

### ISSUE THREAD

**Issue by khaep** [PaddlePaddle/PaddleSeg#1616]
Hello there,

for semantic segmentation purpose, 
I have trained the BiSeNetV2 model with a custom data set.
After finishing the training process, I got reasonable predictions (about 77% mIoU) on the test dataset.
Doing some predictions with [predict.py](https://github.com/PaddlePaddle/PaddleSeg/blob/release/2.3/predict.py) (refer [doc](https://github.com/PaddlePaddle/PaddleSeg/blob/release/2.3/docs/predict/predict.md)) also works fine for the first impressions.

Then, I want to deploy the trained model on the target system. Therefore I read the  [export](https://github-com.translate.goog/PaddlePaddle/PaddleSeg/blob/release/2.3/docs/model_export.md?_x_tr_sl=auto&_x_tr_tl=en&_x_tr_hl=de) instructions and exported the model, so I get the described resulting files.

*  deploy.yaml
*  model.pdiparams
*  model.pdiparams.info
*  model.pdmodel

Now the question is how can I implement the inference of the trained model?
I am looking for something like a python script, that reads an input image, does the inference and returns the predicted image.
I saw that you provide a [script](https://github-com.translate.goog/PaddlePaddle/PaddleSeg/blob/release/2.3/deploy/python/infer.py?_x_tr_sl=auto&_x_tr_tl=en&_x_tr_hl=de) for inference. 
But which lines are necessary for the inference?

Additionaly I thought about converting the model from PaddleSeg to ONNX with [paddle2onnx](https://github-com.translate.goog/PaddlePaddle/Paddle2ONNX/blob/develop/README.md?_x_tr_sl=auto&_x_tr_tl=en&_x_tr_hl=de).
Maybe a custom inference script can look something like this: 

```python
import numpy as np
import cv2
import onnx
import onnxruntime
from onnx import numpy_helper

#set paths
img_path = "/home/PaddleSeg/raw_image.png"
model = "model.onnx"

#read image
img = cv2.imread(img_path)

#normalize the image
img = img.astype(np.float32)/255.0
#img = img - np.array([np.mean(img[:,:,0]), np.mean(img[:,:,1]), np.mean(img[:,:,2])])
#img = img / np.array([np.std(img[:,:,0]), np.std(img[:,:,1]), np.std(img[:,:,2])])

#add fourth dimension to fit input shape of the exported model [-1, 3, -1, -1]
img = np.expand_dims(img, 0)
img = np.transpose(img, (0, 3, 1, 2))

session = onnxruntime.InferenceSession(model, None, providers=["CUDAExecutionProvider"])

input_name = session.get_inputs()[0].name
output_name = session.get_outputs()[0].name

# here the inference is done
result = session.run([output_name], {input_name:np.float32(img)})

# convert the result back to an 320 x 640 image
pred_img = np.uint8(result[0])
pred_img = np.reshape(pred_img, (320, 640))
cv2.imwrite("result.png", pred_img)

```
By using this script, I get some predictions but they differ from the predictions that are created by using the [infer.py](https://github-com.translate.goog/PaddlePaddle/PaddleSeg/blob/release/2.3/deploy/python/infer.py?_x_tr_sl=auto&_x_tr_tl=en&_x_tr_hl=de) script.
So the question is, whats the reason for that?

Is there something wrong with the normalization compared to the run with [infer.py](https://github-com.translate.goog/PaddlePaddle/PaddleSeg/blob/release/2.3/deploy/python/infer.py?_x_tr_sl=auto&_x_tr_tl=en&_x_tr_hl=de)?
Perhaps, the mean and the standard deviation need to be considered at the normalization step.

Here is the content of the config.yml file, that was used for training. Maybe it is useful to answer the questions.
```yml

train_dataset:
  type: Dataset
  dataset_root: dataset
  train_path: dataset/train.txt
  num_classes: 2
  transforms:
    - type: RandomHorizontalFlip
    - type: Normalize
  mode: train

val_dataset:
  type: Dataset
  dataset_root: dataset/FramesForVal/
  val_path: dataset/FramesForVal/val.txt
  num_classes: 2
  transforms:
    - type: Normalize
  mode: val

test_dataset:
  type: Dataset
  dataset_root: dataset/
  test_path: dataset/test.txt
  num_classes: 2
  transforms:
    - type: Normalize
  mode: test

model:
  type: BiSeNetV2
  num_classes: 2

optimizer:
  type: sgd
  weight_decay: 0.0005

loss:
  types:
    - type: CrossEntropyLoss
    - type: CrossEntropyLoss
    - type: CrossEntropyLoss
    - type: CrossEntropyLoss
    - type: CrossEntropyLoss
  coef: [1, 1, 1, 1, 1]

batch_size: 8
iters: 100000

lr_scheduler:
  type: PolynomialDecay
  learning_rate: 0.05
  end_lr: 0.0
  power: 0.9

```
It would be greate if you can give me some answers or hints about that.
I look forward to hearing from you.

Kind regards

**Comment by juncaipeng** [comment]
For the first question, please refer the code in [infer.py](https://github.com/PaddlePaddle/PaddleSeg/blob/release/2.3/deploy/python/infer.py) and the main steps as follows:
* Set PredictConfig
* Create Predictor
* Prepare image
* Run Predictor
* Get results

For the second question, you should convert the image from BGR to RGB by `im = cv2.cvtColor(im, cv2.COLOR_BGR2RGB)`

**Comment by khaep** [comment]
@juncaipeng,
Thank you for the reply.
In the meantime, I modified the [predict.py](https://github-com.translate.goog/PaddlePaddle/PaddleSeg/blob/release/2.3/predict.py?_x_tr_sl=auto&_x_tr_tl=en&_x_tr_hl=de&_x_tr_pto=op) as follows, to get a first inference script to predict single images.

```python
import argparse
import os
import numpy as np
import cv2
import paddle

from paddleseg.cvlibs import manager, Config
from paddleseg.utils import get_sys_env, logger, config_check, get_image_list

from paddleseg import utils
from paddleseg.core import infer
import time

def parse_args():
    parser = argparse.ArgumentParser(description='Model prediction')

    # params of prediction
    parser.add_argument(
        "--config", dest="cfg", help="The config file.", default=None, type=str)
    parser.add_argument(
        '--model_path',
        dest='model_path',
        help='The path of model for prediction',
        type=str,
        default=None)
    parser.add_argument(
        '--image_path',
        dest='image_path',
        help=
        'The path of image, it can be a file or a directory including images',
        type=str,
        default=None)
    parser.add_argument(
        '--save_dir',
        dest='save_dir',
        help='The directory for saving the predicted results',
        type=str,
        default='./output/result')
    return parser.parse_args()

def predict(img, 
            model, 
            model_path,
            transforms, 
            scales=1.0, 
            is_slide=False,
            stride=None,
            crop_size=None):
    
    utils.utils.load_entire_model(model, model_path)
    model.eval()
    
    with paddle.no_grad():
        timer1 = time.time()
        ori_shape = img.shape[:2]
        img, _ = transforms(img)
        img = img[np.newaxis, ...]
        img = paddle.to_tensor(img)

        pred = infer.inference(
                model,
                img,
                ori_shape=ori_shape,
                transforms=transforms.transforms,
                is_slide=is_slide,
                stride=stride,
                crop_size=crop_size)
        pred = paddle.squeeze(pred)
        pred = pred.numpy().astype('uint8')
        timer2 = time.time()
        print("elapsed inf time: ", timer2 - timer1)
        return pred



def main(args):
    env_info = get_sys_env()
    place = 'gpu' if env_info['Paddle compiled with cuda'] and env_info[
        'GPUs used'] else 'cpu'

    paddle.set_device(place)
    cfg = Config(args.cfg)
    img = cv2.imread(args.image_path)

    pred = predict(img, cfg.model, model_path=args.model_path, transforms=cfg.val_dataset.transforms)
    cv2.imwrite('/home/PaddleSeg/InfTest.png', pred)

if __name__ == '__main__':
    args = parse_args()
    main(args)
```
It works and the predicted image looks valid but the inference time on the GPU (measured with the timer variables) is about 447 ms for a single image.  
![Screenshot from 2021-12-15 17-48-48](https://user-images.githubusercontent.com/66534155/146228992-273aeddc-e2ba-44e7-8092-365b9e071dcc.png)
That's to much in my opinion.
Can you give me some tips to improve the speed with manageable effort.
Which adjustments can be made to reduce the inference time?
I know that you can use tensor rt from nvidia to speed up the inference but it is not my main focus at the moment.
Rather I need to do the inference for further development of my system and it is enough if I can achive frame rates between 20 and 60 FPS on a RTX Nvidia GPU.

Kind regards ;)

**Comment by juncaipeng** [comment]
Hi, in order to reduce the inference time, you can export the inference model, and then use the python or cpp api of PaddleInference to deploy the model, such as [infer.py](https://github.com/PaddlePaddle/PaddleSeg/blob/release/2.3/deploy/python/infer.py) and [cpp demo](https://github.com/PaddlePaddle/PaddleSeg/tree/release/2.3/deploy/cpp).  Besides, you have to consider to reduce the pre-process and post-process time.
